src/main/java/co/cask/hydrator/salesforce/plugin/source/batch/SalesforceBatchSource.java

```diff
@@ -0,0 +1,264 @@
+    }
+
+    for (int i = 0; i < fieldNames.length; i++) {
+      String fieldName = fieldNames[i];
+      String value = values[i];
+
+      Schema.Field field = schema.getField(fieldName);
+
+      if (field == null) {
+        continue; // this field is not in schema
+      }
+
+      builder.set(fieldName, convertValue(value, field));
+    }
+
+    emitter.emit(builder.build());
+  }
+
+  /**
+   * Request object for retrieving schema from the Salesforce
+   */
+  class Request {
+    String query;
+    String username;
+    String password;
+    String clientId;
+    String clientSecret;
+    String loginUrl;
+  }
+
+  /**
+   * Get Salesforce schema by query.
+   *
+   * @param request request with credentials and query
+   * @return schema calculated from query
+   * @throws Exception is thrown by httpclientlib
+   */
+  @Path("getSchema")
+  public Schema getSchema(Request request) throws Exception {
+    return SalesforceSchemaUtil.getSchemaFromQuery(new AuthenticatorCredentials(request.username, request.password,
+                                                                                request.clientId, request.clientSecret,
+                                                                                request.loginUrl), request.query);
+  }
+
+  /**
+   * An advanced version of split csv row by comma. We cannot simply split by comma since some values may have comma
+   * inside them and Salesforce does not escape it.
+   *
+   * The format of Salesforce csv row is:
+   * "value1","value2","value3"
+   *
+   * So we are splitting by "\",\"" instead of by comma. Since quotes are not allowed without escaping
+   * in values this works in any case.
+   *
+   *
+   * @param csvRow one row in csv format with quoted values
+   * @return an array of values
+   */
+  private String[] getValuesFromCSVRow(String csvRow) {
+    String[] values = csvRow.split("\",\"");
+
+    values[0] = values[0].substring(1);
+
+    String last = values[values.length - 1];
+    values[values.length - 1] = last.substring(0, last.length() - 1);
+
+    return values;
+  }
+
+  private Object convertValue(String value, Schema.Field field) {
+    Schema fieldSchema = field.getSchema();
+
+    if (fieldSchema.isNullable()) {
+      fieldSchema = fieldSchema.getNonNullable();
+    }
+
+    Schema.Type fieldSchemaType = fieldSchema.getType();
+
+    // empty string is considered null in csv, for all types but string.
+    if (value.isEmpty() && !fieldSchemaType.equals(Schema.Type.STRING)) {
+      return null;
+    }
+
+    Schema.LogicalType logicalType = fieldSchema.getLogicalType();
+    if (fieldSchema.getLogicalType() != null) {
+      switch (logicalType) {
+        case DATE:
+          // date will be in yyyy-mm-dd format
+          return Math.toIntExact(LocalDate.parse(value).toEpochDay());
+        case TIMESTAMP_MILLIS:
+          return Instant.parse(value).toEpochMilli();
+        default:
```

Comment:
Could we ensure we handle all types salesforce supports.

Reply:
This is switch for logicalTypes in cdap schema. I have ensured we support all logical types cdap supports.

Reply:
Talked to Nitin offline will add a doc of types conversions.

Reply:
done